### ISSUE TITLE
i want to submit tcptracer

### ISSUE THREAD

**Issue by jony-one** [iovisor/bcc#3533]
i want to submit tcptracer pr ,is BCC to libbpf conversion 
this is source 
https://github.com/iovisor/bcc/blob/master/tools/tcptracer.py

i will coding ,is ok?

**Comment by chenhengqi** [comment]
@yonghong-song  Is it possible to add a tcp_drop tracepoint to kernel for TCP drop events ?

**Comment by yonghong-song** [comment]
> @yonghong-song Is it possible to add a tcp_drop tracepoint to kernel for TCP drop events ?

There are many reasons for a packet to be dropped. So just adding a tcp_drop tracepoint in the tcp_drop function probably won't fly.
The following is an example for tracepoint to catch checksum errors:
   https://github.com/torvalds/linux/commit/709c0314239992162cba26a860f04319a15860c4
checksum errors will result in packets to be dropped (through tcp_drop) or may simply be simply kfree'ed.

If you can think of other reasons to monitor a specific category of packet drop events, we could propose a new tracepoint.

BTW, I think it is useful to support the above net/tcp_bad_csum in tcpdrop.py with an *explicit* opt-in option like `--csum_error`.

**Comment by chenhengqi** [comment]
The problem is that `tcp_drop` may get inlined (This happens on Ubuntu with kernel 5.10/5.11). 
`sk_drops_add` has the same signature as `tcp_drop`, and the symbol also exists in `/proc/kallsyms`, but it's not traceable. :(

**Comment by jony-one** [comment]
so there are many reasons,now unfriendly at the time of feature development

**Comment by yonghong-song** [comment]
> `sk_drops_add`

Right. `sk_drops_add` is a static inline function hence it is not traceable.

I am aware of `tcp_drop` inline issue and unfortunately we cannot just put a `noinline` attribute in the kernel for kprobe purpose. That is what I suggested earlier. We could develop some tracepoint to track particular categories of drops. csum_error is one of such tracepoints. Other tcp drop categories could be added too.

**Comment by jony-one** [comment]
ok ok.
as  `tracepoint:skb:kfree_skb` ?  can add to .
`kfree_skb` is too simple =)   
For a newbie like me it might be just right ：）

**Comment by yonghong-song** [comment]
No, kfree_skb is not the right kprobe function, the skb to be freed might be a legitimate one.

**Comment by jony-one** [comment]
ok ok 
you can propose a new tracepoint. 
I'm looking forward to it. 
(●'◡'●) happy ...

**Comment by brendangregg** [comment]
I proposed something similar at NetConf 2018: Enhancing the tcp_drop() function to include reason fields for the drop, where `reason` may be an enum or a text string:

https://www.slideshare.net/brendangregg/netconf-2018-bpf-observability/13

Then we can kprobe that function and get the stack trace plus the reason arguments for the drop. Such a function can also include a tracepoint so that the events didn't disappear if inlined.

The idea was generally supported, but I have since not had time to do the work.

The overall steps should be:

1. Ensure tcp_drop() is called in all the right places (and there isn't a kfree_skb() that should be turned into a tcp_drop())
2. Enhance tcp_drop() to include a reason field (and discuss on net-dev/bpf)
3. Add a tracepoint to tcp_drop() that included the reason field
4. Enhance bcc's tcpdrop.py with the extra information (making it a far more useful tool)

Someone needs to start with step 1!

**Comment by yonghong-song** [comment]
@brendangregg Let me take a look at this.

**Comment by jony-one** [comment]
me too

**Comment by yonghong-song** [comment]
@jony-one I planned but got consumed by other activities. If you got time, feel free to work on this. Thanks!

**Comment by jony-one** [comment]
I will submit the patch to the kernel and expect success

**Comment by jony-one** [comment]
so .
> Ensure tcp_drop() is called in all the right places (and there isn't a kfree_skb() that should be turned into a tcp_drop())
Can we skip this step and do it in the revision. Let's start with the second step
> Enhance tcp_drop() to include a reason field (and discuss on net-dev/bpf)
I think enumerations should be distinguished by method, do you have any advice?

**Comment by yonghong-song** [comment]
> so .
> 
> > Ensure tcp_drop() is called in all the right places (and there isn't a kfree_skb() that should be turned into a tcp_drop())
> 
> Can we skip this step and do it in the revision. Let's start with the second step

To get a complete coverage, I think it is worthwhile to check this step as well.

> 
> > Enhance tcp_drop() to include a reason field (and discuss on net-dev/bpf)
> 
> I think enumerations should be distinguished by method, do you have any advice?

Enumerator should be fine.

**Comment by brendangregg** [comment]
What's the thoughts on enumerators?

While it's good coding practice to use enums, debugging/tracing is a little different (e.g., printk()), and it may well be acceptable to have the reason-for-drop field be a char *. That makes it more flexible and easy to trace with bpftrace/bcc.

**Comment by jony-one** [comment]
Let me try

**Comment by brendangregg** [comment]
As it's relevant, I want to link to another comment I made about tracing packet drops: https://github.com/iovisor/bcc/pull/3306#issuecomment-803286723:

> As a tool to investigate packet drops: we shouldn't be making the end user pick through network stack traces to identify what's abnormal if it can be avoided. And it can. I'd have made the tool only print packet drops (filter normal paths), and print the SNMP counter name for the drop followed by the packet details: IP addresses and ports and stack trace, at a minimum. The tool doesn't do that.

So that's where I see this tool heading: With a reason field, we can tell the user which drops are abnormal. Perhaps there could be an additional and optional field for the SNMP couner name as well.

**Comment by brendangregg** [comment]
Just to illustrate, I'd love tcpdrop to turn into:

```
# ./tcpdrop.py
TIME     PID    IP SADDR:SPORT          > DADDR:DPORT          STATE (FLAGS)
20:49:06 0      4  10.32.119.56:443     > 10.66.65.252:22912   CLOSE (ACK)
SNMP                REASON
TCPBacklogDrop      Backlog overflow
	tcp_drop+0x1
	tcp_v4_do_rcv+0x135
	tcp_v4_rcv+0x9c7
	ip_local_deliver_finish+0x62
	ip_local_deliver+0x6f
	ip_rcv_finish+0x129
	ip_rcv+0x28f
	__netif_receive_skb_core+0x432
	__netif_receive_skb+0x18
	netif_receive_skb_internal+0x37
	napi_gro_receive+0xc5
	ena_clean_rx_irq+0x3c3
	ena_io_poll+0x33f
	net_rx_action+0x140
	__softirqentry_text_start+0xdf
	irq_exit+0xb6
	do_IRQ+0x82
	ret_from_intr+0x0
	native_safe_halt+0x6
	default_idle+0x20
	arch_cpu_idle+0x15
	default_idle_call+0x23
	do_idle+0x17f
	cpu_startup_entry+0x73
	rest_init+0xae
	start_kernel+0x4dc
	x86_64_start_reservations+0x24
	x86_64_start_kernel+0x74
	secondary_startup_64+0xa5

[...]
```

So this is just adding two extra fields, SNMP and REASON, where SNMP may be populated or may be blank ("-"). (I also put them on a new line, since I avoid by-default output that is >80 chars wide. This tool could include a `-w` option for wide output, where everything is on one line.)

**Comment by jony-one** [comment]
so cool, feel like there will be a lot of places to use in the future, I have submitted code in the kernel, thank you very much for your advice.
I've reviewed another comment， #3306 ， #1903
Should the relative reason field be set by passing the function directly to the source of the call (simple), or should the exact reason be given by analyzing the context (hard)

**Comment by yonghong-song** [comment]
@jony-one could you cc me the kernel patch so I can be kept in the loop?

**Comment by jony-one** [comment]
> @jony-one could you cc me the kernel patch so I can be kept in the loop?

ok ok

**Comment by chenhengqi** [comment]
I am also working on this issue, but I haven't submitted yet.

@jony-one  could you please also cc me in ?

**Comment by jony-one** [comment]
@yonghong-song  @chenhengqi  @brendangregg  how to get your email to cc

**Comment by chenhengqi** [comment]
hengqi.chen@gmail.com

**Comment by yonghong-song** [comment]
@jony-one you can cc me with yhs@fb.com which is my address used in mailing list.

**Comment by jony-one** [comment]
I submitted it but there is no response is there something wrong, I have cc'd you guys. Can you help check? Thank you!

**Comment by yonghong-song** [comment]
> I submitted it but there is no response is there something wrong, I have cc'd you guys. Can you help check? Thank you!

Thanks for the submission. I replied with some comments.
